### PR TITLE
Admin Place and Remove Tiles

### PR DESCRIPTION
**Pull request by ThatDan123**
CL: [New] Admins can now place and remove tiles

-Placing wires needs looking into as i have no idea how the electrical system needs updating

image

### CODE REVIEW

**Comment by ThatDan123** [comment]
Just gonna ignore all those moth png changes, no idea when they were committed

**Comment by ThatDan123** [comment]
I think the rest of the codacy stuff can be ignored?

File: UnityProject/Assets/Scripts/Player/PlayerNetworkActions.cs [hunk 1]
```diff
@@ -981,6 +986,164 @@ public void ServerAGhost()
 		}
 	}
 
+	[Command]
+	public void CmdPlaceTile(string adminId, string adminToken, int categoryIndex, int tileIndex,
+		Vector3Int worldPosition, int matrixId, OrientationEnum orientation, Color? colour)
+	{
+		var admin = PlayerList.Instance.GetAdmin(adminId, adminToken);
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Player/PlayerNetworkActions.cs]: Curious why you decide to pass an admin ID and token and not find the user by connection as done elsewhere here? E.g. line `966` `if (AdminCommandsManager.IsAdmin(connectionToClient, out _))`

Edit: ah right, 2021 code 🙂

File: UnityProject/Assets/Scripts/UI/Systems/AdminTools/DevTools/GUI_DevTileChanger.cs [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+
+		private void SetUpCategories()
+		{
+			categoryButtonPrefab.SetActive(true);
+
+			for (int i = 0; i < TileCategorySO.Instance.TileCategories.Count; i++)
+			{
+				var newCategory = Instantiate(categoryButtonPrefab, categoryContentPanel.transform);
+
+				var i1 = i;
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/UI/Systems/AdminTools/DevTools/GUI_DevTileChanger.cs:139]: `i1`? The value of `i` should be captured when the listener was created, right?

**Comment by NoooneyDude** [comment]
> Just gonna ignore all those moth png changes, no idea when they were committed

Shouldn't be too hard to undo, right? Just copy the lastest Moth folder to your working dir. Though, up to you 🤷

File: UnityProject/Assets/Scripts/Player/PlayerNetworkActions.cs (hunk 1 above)
Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Player/PlayerNetworkActions.cs]: I could move this to AdminCommandsManager

File: UnityProject/Assets/Scripts/UI/Systems/AdminTools/DevTools/GUI_DevTileChanger.cs (hunk 2 above)
Comment (ThatDan123) [reply UnityProject/Assets/Scripts/UI/Systems/AdminTools/DevTools/GUI_DevTileChanger.cs:139]: image

Rider complains

Comment (NoooneyDude) [reply UnityProject/Assets/Scripts/UI/Systems/AdminTools/DevTools/GUI_DevTileChanger.cs:139]: Oops, you are right 👍

**Review by NoooneyDude — DISMISSED** [review]

**Comment by ThatDan123** [comment]
Should be done now

**Review by NoooneyDude — APPROVED** [review]

**Pull request merged**